src/Controller/Component/CommonComponent.php

```diff
@@ -343,4 +344,22 @@ public function isForeignReferer($ref = null) {
 		return true;
 	}
 
+	/**
+	 * Similar to allowMethod(), but allows only certain extensions.
+	 * Use '' to allow only non-ext.
+	 *
+	 * @param array<string>|string $extensions
+	 * @return void
+	 */
+	public function allowExtension(array|string $extensions): void {
+		$ext = $this->controller->getRequest()->getParam('_ext');
+		if (!$ext && $extensions !== '' || !in_array((string)$ext, (array)$extensions, true)) {
+			throw new NotFoundException(sprintf(
+				'Extension `%s` not allowed, expected `%s`.',
+				$ext,
+				implode(', ', (array)$extensions),
```

Comment:
This could be done using ``Hash::toList($extensions, 'or');`` to slightly improve readability.

https://book.cakephp.org/3/en/core-libraries/text.html#Cake\Utility\Text::toList